services/121-service/src/payments/fsp-integration/safaricom/safaricom.service.ts

```diff
@@ -0,0 +1,241 @@
+      const matchingString =
+        'ed15b1f7-c579-4a59-8224-56520cd3bdf7-payment-1-1636456825275';
+
+      if (matchingString) {
+        const referenceId = matchingString
+          .split('_')[0]
+          .replace('referenceId-', '');
+        const programId = Number(
+          matchingString.split('_')[1].replace('program-', ''),
+        );
+        const payment = Number(
+          matchingString.split('_')[2].replace('payment-', ''),
+        );
+
+        const paTransactionResult = new PaTransactionResultDto();
+        paTransactionResult.fspName = FspName.safaricom;
+        paTransactionResult.referenceId = referenceId;
+        paTransactionResult.status = successStatuses.includes(
+          safaricomRequest.status,
+        )
+          ? StatusEnum.success
+          : StatusEnum.error;
+        paTransactionResult.message = safaricomRequest.status;
+        paTransactionResult.calculatedAmount = Number(safaricomRequest.amount);
+
+        this.transactionsService.storeTransactionUpdateStatus(
+          paTransactionResult,
+          programId,
+          payment,
+        );
+      }
+    }
+  }
+
+  public async processSafaricomResult(
+    safaricomPaymentResultData: any,
+  ): Promise<void> {
+    const safaricomDbRequest = await this.safaricomRequestRepository
+      .createQueryBuilder('safaricom_request')
+      .where('safaricom_request.requestResult ::jsonb @> :requestResult', {
+        requestResult: {
+          ConversationID: safaricomPaymentResultData.Result.ConversationID,
+        },
+      })
+      .getMany();
+
+    const transaction = await this.transactionRepository
+      .createQueryBuilder('transaction')
+      .where('transaction.customData ::jsonb @> :customData', {
+        customData: {
+          requestResult: {
+            ConversationID: safaricomPaymentResultData.Result.ConversationID,
+          },
+        },
+      })
+      .getMany();
+
+    if (
+      safaricomPaymentResultData &&
+      safaricomPaymentResultData.Result &&
+      safaricomPaymentResultData.Result.ResultCode === 0
+    ) {
+      safaricomDbRequest[0].status = StatusEnum.success;
+      transaction[0].status = StatusEnum.success;
+    } else {
+      safaricomDbRequest[0].status = StatusEnum.error;
+      transaction[0].status = StatusEnum.error;
+    }
+
+    safaricomDbRequest[0].paymentResult = safaricomPaymentResultData;
+    const safaricomCustomData = { ...transaction[0].customData };
+    safaricomCustomData['paymentResult'] = safaricomPaymentResultData;
+
+    await this.transactionRepository
+      .createQueryBuilder('transaction')
+      .update('transaction')
+      .set({
+        errorMessage: safaricomPaymentResultData.Result.ResultDesc,
+        status: transaction[0].status,
+        customData: safaricomCustomData,
+      })
+      .where('id = :id', { id: transaction[0].id })
+      .execute();
+
+    await this.safaricomRequestRepository.save(safaricomDbRequest);
```

Comment:
Let's talk this through. I was thinking about the following: 
These queries on both tables will perform badly when there's a lot of data. Because of no indexing. Maybe it would good to add the ConversationID and/or OriginatorConversationID as a separate column with an index. And add the transactionId to the safaricom_request entity.

Reply:
I have added addition properties and One to One relation to Transaction table.